account_invoice_line_sort/models/account_invoice.py

```diff
@@ -0,0 +1,130 @@
+#     as published by the Free Software Foundation, either version 3 of
+#     the License, or (at your option) any later version.
+#
+#     account_invoice_line_sort is distributed in the hope that it will
+#     be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#     GNU Affero General Public License for more details.
+#
+#     You should have received a copy of the
+#     GNU Affero General Public License
+#     along with account_invoice_line_sort.
+#     If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp import models, fields, api
+
+AVAILABLE_SORT_OPTIONS = [
+    ('sequence', 'Sequence'),
+    ('name', 'Description'),
+    ('price_unit', 'Unit Price'),
+    ('price_subtotal', 'Amount'),
+]
+AVAILABLE_ORDER_OPTIONS = [
+    ('asc', 'Ascending'),
+    ('desc', 'Descending')
+]
+
+
+class account_invoice(models.Model):
+    _inherit = "account.invoice"
+    _sort_trigger_fields = ('line_order',
+                            'line_order_direction')
+
+    line_order = fields.Selection(AVAILABLE_SORT_OPTIONS,
+                                  "Sort Lines By",
+                                  default='sequence')
+    line_order_direction = fields.Selection(AVAILABLE_ORDER_OPTIONS,
+                                            "Sort Direction",
+                                            default='asc')
+
+    @api.model
+    def get_partner_sort_options(self, partner_id):
+        res = {}
+        if partner_id:
+            p = self.env['res.partner'].browse(partner_id)
+            res['line_order'] = p.line_order
+            res['line_order_direction'] = p.line_order_direction
+        return res
+
+    @api.multi
+    def onchange_partner_id(self, type, partner_id, date_invoice=False,
+                            payment_term=False, partner_bank_id=False,
+                            company_id=False):
+        res = super(account_invoice,
+                    self).onchange_partner_id(type,
+                                              partner_id,
+                                              date_invoice=date_invoice,
+                                              payment_term=payment_term,
+                                              partner_bank_id=partner_bank_id,
+                                              company_id=company_id)
+        if partner_id:
+            res['value'].update(self.get_partner_sort_options(partner_id))
+        return res
+
+    @api.one
+    def _sort_account_invoice_line(self):
+        line_model = self.env['account.invoice.line']
+        if self.invoice_line:
+            order = "%s %s" % (self.line_order, self.line_order_direction)
+            domain = [('id', 'in', self.invoice_line.ids)]
+            sequence = 0
+            for line in line_model.search(domain, order=order):
```

Comment:
@Cedric-Pigeon e.g. products from 'Food' category before 'Non-Food' products. It's an example. IMO, it would be easier to extend the module for others specific needs if the dot notation is natively supported even if the module doesn't provide this kind of predefined sort option in AVAILABLE_SORT_OPTIONS